### PR TITLE
Refactor: ClientConfigBuilder-specific `BuilderConfig` object for configuration handling

### PR DESCRIPTION
**Pull request by pileks**
This is a continuation of #1465. Huge props to @Niraj-Kamdar for doing the bulk of the work around this.

This PR introduces the concept of `BuilderConfig`, which is a `ClientConfigBuilder`-specific configuration object wherein the individual configuration sections take the shape of a `Record<string, SectionType>`.

This allows us to configure the client in a very readable fashion.
Let's take `config.envs` as an example.
Before, we used to write:

```typescript
const envs = [
  {
    uri: Uri.from("wrap://ens/hello-world.polywrap.eth"),
    env: {
      foo: "bar",
    },
  },
  {
    uri: Uri.from("ens/ethereum.polywrap.eth"),
    env: {
      connection: {
        node: "https://mainnet.infura.io/v3/some_api_key",
        networkNameOrChainId: "mainnet",
      },
    },
  },
];

const config = new ClientConfigBuilder()
  .addDefaults()
  .addEnvs(envs)
  .build();

const client = new PolywrapClient(config);
```

Now we can write a much more succint form of the configuration:
```typescript
const envs = {
      "ens/ethereum.polywrap.eth": {
        connection: {
          networkNameOrChainId: "mainnet",
          node: "https://mainnet.infura.io/v3/some_api_key",
        },
      },
      "ens/hello-world.polywrap.eth": { foo: "bar" },
    };

const config = new ClientConfigBuilder()
  .addDefaults()
  .addEnvs(envs)
  .build();

const client = new PolywrapClient(config);
```

This can also be seen as precursor work to #1283 as it lays the foundation of the "ideal" layout of our final config objects.

Thank you @krisbitney for the note that this PR does a few more things than the description initially explained, these are:

- replace Uri | string with Uri
- remove tracing from core
- fix http server infra build

### CODE REVIEW

File: packages/cli/src/commands/build.ts [hunk 1]
```diff
@@ -174,11 +173,13 @@ async function run(options: Required<BuildCommandOptions>) {
   } = options;
   const logger = createLogger({ verbose, quiet, logFile });
 
-  const envs = await parseWrapperEnvsOption(wrapperEnvs);
+  const envs = await parseWrapperEnvsOption(
+    wrapperEnvs ? wrapperEnvs : undefined
```
Comment (cbrzn) [inline packages/cli/src/commands/build.ts]: nit: what's the reason for adding this ternary if? we can just pass `wrapperEnvs`, right?

**Review by cbrzn — COMMENTED** [review]
thanks for pushing this! looks great :-)

File: packages/cli/src/commands/build.ts (hunk 1 above)
Comment (pileks) [reply packages/cli/src/commands/build.ts]: `wrapperEnvs` in the options is a `WrapperEnvs | false` type (thus making it essentially required, as per @dOrgJelli's CLI-JS PR). `parse` takes `WrapperEnvs | undefined`, so it's just a simple transform.

Comment (Niraj-Kamdar) [reply packages/cli/src/commands/build.ts]: Then parse should take `WrapperEnvs  | undefined | bool` I guess and do proper transformation inside the function.

Comment (pileks) [reply packages/cli/src/commands/build.ts]: Done!

File: packages/js/client-config-builder/src/bundles/getDefaultConfig.ts [hunk 2]
```diff
@@ -25,147 +26,73 @@ export const defaultWrappers = {
   sha3: "wrap://ens/goerli/sha3.wrappers.eth",
   uts46: "wrap://ens/goerli/uts46-lite.wrappers.eth",
   graphNode: "wrap://ens/goerli/graph-node.wrappers.eth",
-  ensTextRecordResolver:
-    "wrap://ipfs/QmfRCVA1MSAjUbrXXjya4xA9QHkbWeiKRsT7Um1cvrR7FY",
+  concurrentInterface: "wrap://ens/goerli/interface.concurrent.wrappers.eth",
 };
 
-export const defaultPackages = {
```
Comment (krisbitney) [inline packages/js/client-config-builder/src/bundles/getDefaultConfig.ts]: Why are we removing these constants? We should instead be using them everywhere instead of string literals.

File: packages/js/client-config-builder/src/types/IClientConfigBuilder.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
+import { BuilderConfig } from "./configs/BuilderConfig";
+import { ClientConfig } from "./configs/ClientConfig";
+
+import { CoreClientConfig, Wrapper, IWrapPackage } from "@polywrap/core-js";
+import { UriResolverLike } from "@polywrap/uri-resolvers-js";
+
+export type TUri = string;
```
Comment (krisbitney) [inline packages/js/client-config-builder/src/types/IClientConfigBuilder.ts]: The names TUri and TEnv could be misleading since they are aliases for specific types but names like generic types. I think we should instead just use the actual types (e.g. `string`) directly.

File: packages/js/client-config-builder/src/types/configs/PolywrapClientConfig.ts [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  IUriRedirect,
+  IUriWrapper,
+  UriResolverLike,
+} from "./types";
+
+import { IWrapperCache } from "@polywrap/uri-resolvers-js";
+import { TracerConfig } from "@polywrap/tracing-js";
+import { Uri } from "@polywrap/core-js";
+
+export interface PolywrapClientConfig<TUri extends Uri | string = string> {
```
Comment (krisbitney) [inline packages/js/client-config-builder/src/types/configs/PolywrapClientConfig.ts]: I think this belongs with the PolywrapClient, not with the config builder. What was the thought process behind the move?

File: packages/js/client-config-builder/src/types/configs/PolywrapCoreClientConfig.ts [hunk 5]
```diff
@@ -0,0 +1,15 @@
+import { Env, InterfaceImplementations } from "./types";
+
+import { IUriResolver, Uri } from "@polywrap/core-js";
+import { TracerConfig } from "@polywrap/tracing-js";
+import { IWrapperCache } from "@polywrap/uri-resolvers-js";
+
+export interface PolywrapCoreClientConfig<
```
Comment (krisbitney) [inline packages/js/client-config-builder/src/types/configs/PolywrapCoreClientConfig.ts]: I think this belongs with the core client, not with the config builder. What was the thought process behind the move?

File: packages/js/client/src/PolywrapClient.ts [hunk 6]
```diff
@@ -1,36 +1,277 @@
...
+import {
+  compareSignature,
+  DeserializeManifestOptions,
+  ImportedModuleDefinition,
+  WrapManifest,
+} from "@polywrap/wrap-manifest-types-js";
+import { Tracer, TracerConfig } from "@polywrap/tracing-js";
 
-export class PolywrapClient extends PolywrapCoreClient {
+export class PolywrapClient<
```
Comment (krisbitney) [inline packages/js/client/src/PolywrapClient.ts:35]: Please add doc comments to the methods in PolywrapClient and update the readme snippets

File: packages/js/core-client/src/PolywrapCoreClient.ts [hunk 7]
```diff
@@ -18,72 +16,35 @@ import {
...
   WrapErrorCode,
 } from "@polywrap/core-js";
 import { msgpackEncode, msgpackDecode } from "@polywrap/msgpack-js";
 import {
-  compareSignature,
   DeserializeManifestOptions,
   WrapManifest,
-  ImportedModuleDefinition,
 } from "@polywrap/wrap-manifest-types-js";
-import { Tracer, TracerConfig, TracingLevel } from "@polywrap/tracing-js";
```
Comment (krisbitney) [inline packages/js/core-client/src/PolywrapCoreClient.ts:33]: While I think it might be correct to removing tracing from the core client and core, I think it should be a separate PR so it can be evaluated in isolation

**Review by krisbitney — DISMISSED** [review]
There is a lot of cool stuff happening in this PR. Thanks for the amazing work!

There are five things happening in this PR:

1. add `BuilderConfig` and changes to `client-config-builder` UX
2. replace `Uri | string` with `Uri`
3. remove tracing from core
4. remove constants from default config
5. fix http server infra build

In the future, I think it would be better to try to do separate PRs. It's easier to review smaller PRs, and fewer issues will be missed by reviewers.

But again, there is a ton of value here. I'm excited to see this get merged!

File: packages/js/client-config-builder/src/bundles/getDefaultConfig.ts (hunk 2 above)
Comment (krisbitney) [reply packages/js/client-config-builder/src/bundles/getDefaultConfig.ts]: The constants solve a real user problem. Before removing them, I'd like to see an alternative solution to the problem.

**Comment by krisbitney** [comment]
If tracing is only implemented in `PolywrapClient`, how does that affect the usefullness of the tracing? If it's no longer effective, should we discuss removing tracing altogether?

File: packages/js/client-config-builder/src/bundles/getDefaultConfig.ts (hunk 2 above)
Comment (pileks) [reply packages/js/client-config-builder/src/bundles/getDefaultConfig.ts]: Agreed, and there is a non-destructive path here, so I took it.

File: packages/js/client-config-builder/src/types/IClientConfigBuilder.ts (hunk 3 above)
Comment (pileks) [reply packages/js/client-config-builder/src/types/IClientConfigBuilder.ts]: Agreed!

File: packages/js/client-config-builder/src/bundles/getDefaultConfig.ts (hunk 2 above)
Comment (krisbitney) [reply packages/js/client-config-builder/src/bundles/getDefaultConfig.ts]: Awesome. Thanks!!

File: packages/js/client-config-builder/src/types/configs/PolywrapClientConfig.ts (hunk 4 above)
Comment (krisbitney) [reply packages/js/client-config-builder/src/types/configs/PolywrapClientConfig.ts]: To clarify, I'm not saying the decision is wrong at all. I don't know. I'm just wondering what the benefit of placing it here is, and if there is a cost.

File: packages/js/client-config-builder/src/types/configs/PolywrapCoreClientConfig.ts (hunk 5 above)
Comment (krisbitney) [reply packages/js/client-config-builder/src/types/configs/PolywrapCoreClientConfig.ts]: Same as I said above: To clarify, I'm not saying the decision is wrong at all. I don't know.

File: packages/js/client/src/PolywrapClient.ts (hunk 6 above)
Comment (krisbitney) [reply packages/js/client/src/PolywrapClient.ts:35]: I know this is a pain but I think you can just copy/paste them from the PolywrapCoreClient

File: packages/js/client-config-builder/src/types/configs/PolywrapClientConfig.ts (hunk 4 above)
Comment (pileks) [reply packages/js/client-config-builder/src/types/configs/PolywrapClientConfig.ts]: https://github.com/polywrap/toolchain/pull/1498

This PR tackles these issues, as I completely agree with what you're saying here.

`ClientConfigBuilder` should really only concern itself with outputting a single `CoreClientConfig`, and the Client should accept that as its primary source of config. The other config objects should only be handled by the appropriate client.

Comment (krisbitney) [reply packages/js/client-config-builder/src/types/configs/PolywrapClientConfig.ts]: Super cool! I'm excited about these changes. You and Niraj did some amazing work.

File: packages/js/client-config-builder/src/types/IClientConfigBuilder.ts [hunk 8]
```diff
@@ -1,43 +1,27 @@
...
-
-  // $start: IClientConfigBuilder-buildCoreConfig
   /**
    * Build a sanitized core client configuration that can be passed to the PolywrapClient or PolywrapCoreClient constructors
    *
-   * @returns CoreClientConfig<Uri> that results from applying all the steps in the builder pipeline
+   * @returns CoreClientConfig that results from applying all the steps in the builder pipeline
    */
-  buildCoreConfig(): CoreClientConfig<Uri>;
+  build(): CoreClientConfig;
```
Comment (nerfZael) [inline packages/js/client-config-builder/src/types/IClientConfigBuilder.ts:15]: `IClientConfigBuilder` seems to be missing a getter for config

File: packages/js/client-config-builder/src/BaseClientConfigBuilder.ts [hunk 9]
```diff
@@ -1,54 +1,60 @@
...
+    redirects: {},
+    wrappers: {},
+    packages: {},
     resolvers: [],
   };
 
-  abstract addDefaults(): ClientConfigBuilder;
-  abstract buildCoreConfig(): CoreClientConfig<Uri>;
+  abstract addDefaults(): IClientConfigBuilder;
+  abstract build(
```
Comment (nerfZael) [inline packages/js/client-config-builder/src/BaseClientConfigBuilder.ts]: I don't think the `BaseClientConfigBuilder` should be aware of a wrapperCache

**Review by nerfZael — COMMENTED** [review]
Most of this looks great.
I left two comments.
Additionally, I find the `Generic**` types really awkward

**Comment by pileks** [comment]
> Most of this looks great. I left two comments. Additionally, I find the `Generic**` types really awkward

We will delete those in a future PR:
https://github.com/polywrap/toolchain/issues/1506

File: packages/js/client-config-builder/package.json [hunk 10]
```diff
@@ -32,6 +32,7 @@
     "@polywrap/ipfs-plugin-js": "0.10.0-pre.7",
     "@polywrap/ipfs-resolver-plugin-js": "0.10.0-pre.7",
     "@polywrap/logger-plugin-js": "0.10.0",
+    "@polywrap/tracing-js": "0.10.0-pre.7",
```
Comment (cbrzn) [inline packages/js/client-config-builder/package.json]: this is not being used in this package

**Review by cbrzn — COMMENTED** [review]
looking solid :clap: - i share the sentiment of Jure of having the two types but I think it's fine to fix this in another PR so we don't get things bigger. 

The only idea I would give is that, since we want to fix the `TUri` generic not going in through the entire client, we maybe have this generic only in the builder, and this takes care of sanitizing the URIs, the client would only expect an URI and wont know about any generics. Making the core types just expect `Uri` instead of `TUri` 

Just one minor thing to change (an unused dependency) and we're good to go

**Comment by Niraj-Kamdar** [comment]
> If tracing is only implemented in `PolywrapClient`, how does that affect the usefullness of the tracing? If it's no longer effective, should we discuss removing tracing altogether?

It will still be beneficial to trace subinvocation traces, tryResolverUri, loadWrapper, or any other top-level functions.

**Comment by cbrzn** [comment]
> > If tracing is only implemented in `PolywrapClient`, how does that affect the usefullness of the tracing? If it's no longer effective, should we discuss removing tracing altogether?
> 
> It will still be beneficial to trace subinvocation traces, tryResolverUri, loadWrapper, or any other top-level functions.

tbh i feel we're not really using the tracing - i think discussing to improve/remove it is worth it

File: packages/js/core/src/types/Uri.ts [hunk 11]
```diff
@@ -194,4 +179,16 @@ export class Uri {
       throw Error(`Unknown uri type, cannot convert. ${JSON.stringify(uri)}`);
     }
   }
+
+  // $start: Uri-toString
+  /** @returns Uri string representation */
+  public toString(): string /* $ */ {
```
Comment (Niraj-Kamdar) [inline packages/js/core/src/types/Uri.ts:185]: Why do we need this?

**Review by Niraj-Kamdar — APPROVED** [review]
LGTM! Let's get this merged and then cleanup legacy

**Review by cbrzn — APPROVED** [review]

File: packages/js/core/src/types/Uri.ts (hunk 11 above)
Comment (pileks) [reply packages/js/core/src/types/Uri.ts:185]: I'll assume that this would render the URI as a string within interpolated strings, but that's as far as I can fathom this.

TBH I've only now seen this and it's something that's been in the Uri class before, it's just been moved from lines 104-115 to here.

Comment (Niraj-Kamdar) [reply packages/js/core/src/types/Uri.ts:185]: Yeah I noticed that but not sure we need this functions in the first place
